File: android-sdk-unity/src/main/java/com/appboy/unity/EventSubscriberFactory.java
package com.appboy.unity;

import com.appboy.events.ContentCardsUpdatedEvent;
import com.appboy.events.FeedUpdatedEvent;
import com.appboy.events.IEventSubscriber;
import com.appboy.events.InAppMessageEvent;
import com.appboy.unity.configuration.UnityConfigurationProvider;
import com.appboy.unity.utils.MessagingUtils;
import com.braze.support.BrazeLogger;

public class EventSubscriberFactory {
  private static final String TAG = BrazeLogger.getBrazeLogTag(EventSubscriberFactory.class);

  public static IEventSubscriber<InAppMessageEvent> createInAppMessageEventSubscriber(final UnityConfigurationProvider unityConfigurationProvider) {
    return inAppMessageEvent -> {
      String unityGameObjectName = unityConfigurationProvider.getInAppMessageListenerGameObjectName();
      String unityCallbackFunctionName = unityConfigurationProvider.getInAppMessageListenerCallbackMethodName();
      boolean isInAppMessageEventSent = MessagingUtils.sendInAppMessageReceivedMessage(unityGameObjectName, unityCallbackFunctionName, inAppMessageEvent.getInAppMessage());
      BrazeLogger.d(TAG, (isInAppMessageEventSent ? "Successfully sent" : "Failure to send") + " in-app message event to Unity Player");
    };
  }

  public static IEventSubscriber<FeedUpdatedEvent> createFeedUpdatedEventSubscriber(final UnityConfigurationProvider unityConfigurationProvider) {
    return feedUpdatedEvent -> {
      String unityGameObjectName = unityConfigurationProvider.getFeedListenerGameObjectName();
      String unityCallbackFunctionName = unityConfigurationProvider.getFeedListenerCallbackMethodName();
      boolean isFeedUpdatedEventSent = MessagingUtils.sendFeedUpdatedEventToUnity(unityGameObjectName, unityCallbackFunctionName, feedUpdatedEvent);
      BrazeLogger.d(TAG, (isFeedUpdatedEventSent ? "Successfully sent" : "Failure to send") + " Feed updated event to Unity Player");
    };
  }

  public static IEventSubscriber<ContentCardsUpdatedEvent> createContentCardsEventSubscriber(final UnityConfigurationProvider unityConfigurationProvider) {
    return contentCardsUpdatedEvent -> {
      String unityGameObjectName = unityConfigurationProvider.getContentCardsUpdatedListenerGameObjectName();
      String unityCallbackFunctionName = unityConfigurationProvider.getContentCardsUpdatedListenerCallbackMethodName();
      boolean isContentCardsEventSent = MessagingUtils.sendContentCardsUpdatedEventToUnity(unityGameObjectName, unityCallbackFunctionName, contentCardsUpdatedEvent);
      BrazeLogger.d(TAG, (isContentCardsEventSent ? "Successfully sent" : "Failure to send") + " Content Cards updated event to Unity Player");
    };
  }
}

File: android-sdk-unity/src/main/java/com/appboy/unity/EventSubscriberFactory.kt
package com.appboy.unity

import com.appboy.events.FeedUpdatedEvent
import com.appboy.events.IEventSubscriber
import com.appboy.unity.configuration.UnityConfigurationProvider
import com.appboy.unity.utils.MessagingUtils.sendContentCardsUpdatedEventToUnity
import com.appboy.unity.utils.MessagingUtils.sendFeedUpdatedEventToUnity
import com.appboy.unity.utils.MessagingUtils.sendInAppMessageReceivedMessage
import com.braze.events.ContentCardsUpdatedEvent
import com.braze.events.InAppMessageEvent
import com.braze.support.BrazeLogger.brazelog
import com.braze.support.BrazeLogger.getBrazeLogTag

object EventSubscriberFactory {
    private val TAG = getBrazeLogTag(EventSubscriberFactory::class.java)

    fun createInAppMessageEventSubscriber(config: UnityConfigurationProvider): IEventSubscriber<InAppMessageEvent> {
        return IEventSubscriber { inAppMessageEvent: InAppMessageEvent ->
            val isInAppMessageEventSent =
                sendInAppMessageReceivedMessage(
                    config.inAppMessageListenerGameObjectName,
                    config.inAppMessageListenerCallbackMethodName,
                    inAppMessageEvent.inAppMessage
                )
            brazelog(TAG) { "Did send in-app message event to Unity Player?: $isInAppMessageEventSent" }
        }
    }

    fun createFeedUpdatedEventSubscriber(config: UnityConfigurationProvider): IEventSubscriber<FeedUpdatedEvent> {
        return IEventSubscriber { feedUpdatedEvent: FeedUpdatedEvent ->
            val isFeedUpdatedEventSent = sendFeedUpdatedEventToUnity(
                config.feedListenerGameObjectName,
                config.feedListenerCallbackMethodName,
                feedUpdatedEvent
            )
            brazelog(TAG) { "Did send Feed updated event to Unity Player?: $isFeedUpdatedEventSent" }
        }
    }

    fun createContentCardsEventSubscriber(config: UnityConfigurationProvider): IEventSubscriber<ContentCardsUpdatedEvent> {
        return IEventSubscriber { contentCardsUpdatedEvent: ContentCardsUpdatedEvent ->
            val isContentCardsEventSent =
                sendContentCardsUpdatedEventToUnity(
                    config.contentCardsUpdatedListenerGameObjectName,
                    config.contentCardsUpdatedListenerCallbackMethodName,
                    contentCardsUpdatedEvent
                )
            brazelog(TAG) { "Did send Content Cards updated event to Unity Player?: $isContentCardsEventSent" }
        }
    }
}
